File: simittag/pose.py
"""
Pose of a circle from its perspective projection.

A circle viewed by a pinhole camera projects to a conic. The rays through the
image conic form a cone in normalized camera coordinates, and asking which
planes cut that cone in a circle recovers the circle's supporting plane, up to
the inherent 2-fold ambiguity (two mirror planes; the data decode picks one).
This is the classical bias-free method for circular-feature pose; see
Y. C. Shiu and S. Ahmad, "3D location of circular and spherical features by
monocular model-based vision," IEEE SMC 1989, and K. Kanatani and W. Liu,
"3D interpretation of conics and orthogonality," CVGIP 1993. Circular fiducial
systems have used it since Rice, Beresford and Harle's Cantag (PerCom 2006).

Derivation used here: scale the conic matrix so its eigenvalues satisfy
l1 >= l2 > 0 > l3. In the eigenbasis the cone is elliptical, fattest along v1
and narrowest along v3. A plane cuts it in a circle when tilted about v2 by
the angle theta with cos(theta) = sqrt((l2-l3)/(l1-l3)) and sin(theta) =
sqrt((l1-l2)/(l1-l3)); the two tilt directions are the two solutions. For the
circle of unit radius, the center sits at inverse distance rho =
sqrt(-l1*l3)/l2 along the tilted axis, displaced u = sqrt((l1-l2)(l2-l3))/l2
within the plane. The homography columns are then the plane's in-plane basis
and the center, so marker-plane points (X, Y, 1) map through it to rays, and
through K to pixels. Sampling the data rings through this homography is true
perspective, which an affine model cannot provide: concentric circles do not
project to concentric ellipses.
"""
from __future__ import annotations
import numpy as np


def ellipse_to_conic(cx, cy, MA, ma, angle_deg):
    """cv2.fitEllipse geometric form -> 3x3 conic matrix C with [x y 1] C [x y 1]^T = 0."""
    a, b = MA / 2.0, ma / 2.0
    th = np.radians(angle_deg)
    R = np.array([[np.cos(th), -np.sin(th)], [np.sin(th), np.cos(th)]])
    M = R @ np.diag([1.0 / a**2, 1.0 / b**2]) @ R.T
    c = np.array([cx, cy])
    C = np.zeros((3, 3))
    C[:2, :2] = M
    C[:2, 2] = -M @ c
    C[2, :2] = -M @ c
    C[2, 2] = c @ M @ c - 1.0
    return C


def circle_poses_from_conic(C):
    """
    Conic (normalized camera coords) -> the two 3x3 homographies mapping the
    unit circle's plane into rays. Columns: plane x basis, plane y basis,
    circle center. Derivation in the module docstring.
    """
    w, V = np.linalg.eigh(C)          # symmetric -> real eigenpairs (ascending)
    if np.sum(w < 0) > 1:             # conic defined up to scale; fix signature +,+,-
        w, V = -w, V
    idx = np.argsort(w)[::-1]         # descending: l1 >= l2 > 0 > l3
    w, V = w[idx], V[:, idx]
    if V[2, 2] < 0:                   # plane normal points toward the camera
        V[:, 2] = -V[:, 2]
    j = int(np.argmax(np.abs(V[:, 1])))
    if V[j, 1] < 0:                   # eigenvector signs are solver-arbitrary; pin
        V[:, 1] = -V[:, 1]            # col 1 so any implementation (LAPACK, the Rust
                                      # port's 3x3 solver) lands on the SAME H rather
                                      # than one rotated 180 deg in-plane
    if np.linalg.det(V) < 0:          # right-handed basis
        V[:, 0] = -V[:, 0]
    l1, l2, l3 = w

    span = l3 - l1
    ct = np.sqrt(max(0.0, (l3 - l2) / span))            # cos(tilt) in the eigenbasis
    st = np.sqrt(max(0.0, (l2 - l1) / span))            # sin(tilt)
    u = np.sqrt(max(0.0, (l2 - l1) * (l3 - l2))) / l2   # center's in-plane offset
    rho = np.sqrt(max(0.0, -l1 * l3 / (l2 * l2)))       # inverse center distance
    v1, v2, v3 = V[:, 0], V[:, 1], V[:, 2]

    out = []
    for s in (+1.0, -1.0):
        b1 = ct * v1 + s * st * v3        # plane x basis
        axis = -s * st * v1 + ct * v3     # tilted axis through the center
        center = (s * u * b1 + axis) / rho
        out.append(np.column_stack([b1, v2, center]))
    return out


def pose_homographies(ellipse_geom, K):
    """
    ellipse_geom = ((cx,cy),(MA,ma),angle) from cv2.fitEllipse (pixel coords).
    Return up to 2 homographies H (3x3) mapping marker-plane (X,Y,1) -> image pixels,
    where the unit circle (X^2+Y^2=1) is the fitted ellipse.
    """
    (cx, cy), (MA, ma), ang = ellipse_geom
    C_pix = ellipse_to_conic(cx, cy, MA, ma, ang)
    C_norm = K.T @ C_pix @ K                  # to normalized camera coords
    Hs = []
    for H_norm in circle_poses_from_conic(C_norm):
        H_pix = K @ H_norm
        if abs(H_pix[2, 2]) > 1e-12:
            H_pix = H_pix / H_pix[2, 2]
        Hs.append(H_pix)
    return Hs


def apply_H(H, X, Y):
    p = H @ np.array([X, Y, 1.0])
    return p[0] / p[2], p[1] / p[2]


def decompose_H(H, K):
    """
    Homography (marker-plane unit circle -> image px) -> (R, t) of the marker in
    CAMERA coordinates, with the marker's z axis = its surface normal. Standard
    K^-1 H decomposition with orthonormalization. Scale is in 'unit-circle radii'
    (the fitted outer ellipse = radius 1), i.e. metric up to marker size.
    """
    Kinv = np.linalg.inv(K)
    L = Kinv @ H
    l1 = np.linalg.norm(L[:, 0])
    l2 = np.linalg.norm(L[:, 1])
    lam = 2.0 / (l1 + l2)
    L = L * lam
    r1 = L[:, 0]; r2 = L[:, 1]; t = L[:, 2]
    r1 /= (np.linalg.norm(r1) or 1.0)
    r2 = r2 - r1 * (r1 @ r2)
    r2 /= (np.linalg.norm(r2) or 1.0)
    r3 = np.cross(r1, r2)
    R = np.column_stack([r1, r2, r3])
    if t[2] < 0:                 # marker must be in front of camera
        R[:, :2] *= -1; t = -t
        R[:, 2] = np.cross(R[:, 0], R[:, 1])
    return R, t


def tilt_from_H(H, K):
    """Approximate tilt angle (deg) of the marker plane normal vs camera axis."""
    Kinv = np.linalg.inv(K)
    h1, h2 = Kinv @ H[:, 0], Kinv @ H[:, 1]
    n1, n2 = h1 / np.linalg.norm(h1), h2 / np.linalg.norm(h2)
    normal = np.cross(n1, n2)
    normal /= np.linalg.norm(normal)
    return float(np.degrees(np.arccos(min(1.0, abs(normal[2])))))

File: rust/simittag-core/src/pose.rs
//! Conic -> perspective transform. Port of simittag/pose.py (itself a port of
//! Cantag's TransformEllipseFull). Gated at 1e-9 against fixtures/geometry.json.
//!
//! The eigenvector sign canonicalization matches pose.py exactly (col 2: normal
//! toward camera; col 1: largest-|component| positive; col 0: det=+1) -- that
//! rule exists precisely so this port and LAPACK land on the SAME homography.

use crate::mat::{self, M3, V3};

#[derive(Debug, Clone, Copy)]
pub struct EllipseGeom {
    pub cx: f64,
    pub cy: f64,
    pub major: f64, // full axis, cv2.fitEllipse convention (either may be larger)
    pub minor: f64,
    pub angle_deg: f64,
}

/// cv2.fitEllipse geometric form -> 3x3 conic C with [x y 1] C [x y 1]^T = 0.
pub fn ellipse_to_conic(g: &EllipseGeom) -> M3 {
    let a = g.major / 2.0;
    let b = g.minor / 2.0;
    let th = g.angle_deg.to_radians();
    let (s, c) = th.sin_cos();
    // M = R diag(1/a^2, 1/b^2) R^T
    let ia = 1.0 / (a * a);
    let ib = 1.0 / (b * b);
    let m00 = c * c * ia + s * s * ib;
    let m01 = c * s * ia - s * c * ib;
    let m11 = s * s * ia + c * c * ib;
    let mc = [
        m00 * g.cx + m01 * g.cy,
        m01 * g.cx + m11 * g.cy,
    ];
    [
        [m00, m01, -mc[0]],
        [m01, m11, -mc[1]],
        [
            -mc[0],
            -mc[1],
            g.cx * mc[0] + g.cy * mc[1] - 1.0,
        ],
    ]
}

/// Conic (normalized camera coords) -> two 3x4-equivalent transforms, returned
/// as the two homography column triples (cols 0,1,3 of the 4x4), matching
/// Python's H_norm = T[:3][:, [0,1,3]].
fn transforms_from_conic(c: &M3) -> [M3; 2] {
    let (mut w, mut v) = mat::eigh3(c);
    if w.iter().filter(|&&x| x < 0.0).count() > 1 {
        for x in &mut w {
            *x = -*x;
        }
    }
    // sort descending: l1 >= l2 >= l3
    let mut idx = [0usize, 1, 2];
    idx.sort_by(|&i, &j| w[j].partial_cmp(&w[i]).unwrap());
    let ws = [w[idx[0]], w[idx[1]], w[idx[2]]];
    let mut vs = [[0.0f64; 3]; 3];
    for (col, &i) in idx.iter().enumerate() {
        for row in 0..3 {
            vs[row][col] = v[row][i];
        }
    }
    v = vs;
    if v[2][2] < 0.0 {
        for row in &mut v {
            row[2] = -row[2];
        }
    }
    let mut j = 0;
    for r in 1..3 {
        if v[r][1].abs() > v[j][1].abs() {
            j = r;
        }
    }
    if v[j][1] < 0.0 {
        for row in &mut v {
            row[1] = -row[1];
        }
    }
    if mat::det3(&v) < 0.0 {
        for row in &mut v {
            row[0] = -row[0];
        }
    }
    let (l1, l2, l3) = (ws[0], ws[1], ws[2]);

    let denom = l3 - l1;
    let pmcos = ((l3 - l2) / denom).max(0.0).sqrt();
    let pmsin = ((l2 - l1) / denom).max(0.0).sqrt();
    let tx = ((l2 - l1) * (l3 - l2)).max(0.0).sqrt() / l2;
    let scale = (-l1 * l3 / (l2 * l2)).max(0.0).sqrt(); // bullseye_size = 1

    let mut out = [[[0.0f64; 3]; 3]; 2];
    for (oi, sgn) in [(0usize, 1.0f64), (1, -1.0)] {
        // R1 @ r2 @ trans, keeping only columns 0, 1, 3 of the 4x4 product.
        // r2 rotates about y: [[pmcos,0,-sgn*pmsin],[0,1,0],[sgn*pmsin,0,pmcos]]
        // trans: x += sgn*tx/scale, z += 1/scale.
        let r2 = [
            [pmcos, 0.0, -sgn * pmsin],
            [0.0, 1.0, 0.0],
            [sgn * pmsin, 0.0, pmcos],
        ];
        let r12 = mat::matmul(&v, &r2);
        let tcol = [sgn * tx / scale, 0.0, 1.0 / scale];
        let tcam = mat::matvec(&r12, &tcol);
        out[oi] = [
            [r12[0][0], r12[0][1], tcam[0]],
            [r12[1][0], r12[1][1], tcam[1]],
            [r12[2][0], r12[2][1], tcam[2]],
        ];
    }
    out
}

/// Ellipse (pixel coords) + K -> up to 2 homographies mapping marker-plane
/// (X, Y, 1) -> image pixels, unit circle = the fitted ellipse.
pub fn pose_homographies(geom: &EllipseGeom, k: &M3) -> Vec<M3> {
    let c_pix = ellipse_to_conic(geom);
    let c_norm = mat::matmul(&mat::matmul(&mat::transpose(k), &c_pix), k);
    let mut out = Vec::with_capacity(2);
    for h_norm in transforms_from_conic(&c_norm) {
        let mut h_pix = mat::matmul(k, &h_norm);
        if h_pix[2][2].abs() > 1e-12 {
            let s = 1.0 / h_pix[2][2];
            for row in &mut h_pix {
                for e in row {
                    *e *= s;
                }
            }
        }
        out.push(h_pix);
    }
    out
}

pub fn apply_h(h: &M3, x: f64, y: f64) -> (f64, f64) {
    let p = mat::matvec(h, &[x, y, 1.0]);
    (p[0] / p[2], p[1] / p[2])
}

/// Homography -> (R, t) in camera coordinates; scale in unit-circle radii.
pub fn decompose_h(h: &M3, k: &M3) -> (M3, V3) {
    let kinv = mat::inv3(k);
    let mut l = mat::matmul(&kinv, h);
    let c0 = [l[0][0], l[1][0], l[2][0]];
    let c1 = [l[0][1], l[1][1], l[2][1]];
    let lam = 2.0 / (mat::norm(&c0) + mat::norm(&c1));
    for row in &mut l {
        for e in row {
            *e *= lam;
        }
    }
    let mut r1 = [l[0][0], l[1][0], l[2][0]];
    let mut r2 = [l[0][1], l[1][1], l[2][1]];
    let mut t = [l[0][2], l[1][2], l[2][2]];
    let n1 = mat::norm(&r1);
    let n1 = if n1 != 0.0 { n1 } else { 1.0 }; // Python: `or 1.0`
    for e in &mut r1 {
        *e /= n1;
    }
    let d = mat::dot(&r1, &r2);
    for i in 0..3 {
        r2[i] -= r1[i] * d;
    }
    let n2 = mat::norm(&r2);
    let n2 = if n2 != 0.0 { n2 } else { 1.0 };
    for e in &mut r2 {
        *e /= n2;
    }
    let mut r3 = mat::cross(&r1, &r2);
    if t[2] < 0.0 {
        for i in 0..3 {
            r1[i] = -r1[i];
            r2[i] = -r2[i];
            t[i] = -t[i];
        }
        r3 = mat::cross(&r1, &r2);
    }
    let r = [
        [r1[0], r2[0], r3[0]],
        [r1[1], r2[1], r3[1]],
        [r1[2], r2[2], r3[2]],
    ];
    (r, t)
}

/// Approximate tilt (deg) of the marker plane normal vs the camera axis.
pub fn tilt_from_h(h: &M3, k: &M3) -> f64 {
    let kinv = mat::inv3(k);
    let h1 = mat::matvec(&kinv, &[h[0][0], h[1][0], h[2][0]]);
    let h2 = mat::matvec(&kinv, &[h[0][1], h[1][1], h[2][1]]);
    let n1n = mat::norm(&h1);
    let n2n = mat::norm(&h2);
    let n1 = [h1[0] / n1n, h1[1] / n1n, h1[2] / n1n];
    let n2 = [h2[0] / n2n, h2[1] / n2n, h2[2] / n2n];
    let mut nrm = mat::cross(&n1, &n2);
    let nn = mat::norm(&nrm);
    for e in &mut nrm {
        *e /= nn;
    }
    nrm[2].abs().min(1.0).acos().to_degrees()
}
